internal/discovery/foreign-cluster-operator/auth.go

```diff
@@ -0,0 +1,204 @@
+package foreign_cluster_operator
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+	discoveryv1alpha1 "github.com/liqotech/liqo/apis/discovery/v1alpha1"
+	"github.com/liqotech/liqo/pkg/auth"
+	"github.com/liqotech/liqo/pkg/crdClient"
+	"github.com/liqotech/liqo/pkg/kubeconfig"
+	"io/ioutil"
+	v1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	client_scheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog"
+	"net/http"
+	"strings"
+)
+
+// get a client to the remote cluster
+// if the ForeignCluster has a reference to the secret's role, load the configurations from that secret
+// else try to get a role from the remote cluster
+//
+// first of all, if the status is pending we can try to get a role with an empty token, if the remote cluster allows it
+// the status will become accepted
+//
+// if our status is EmptyRefused, this means that the remote cluster refused out request with the empty token, so we will
+// wait to have a token to ask for the role again
+//
+// while we are waiting for that secret this function will return no error, but an empty client
+func (r *ForeignClusterReconciler) getRemoteClient(fc *discoveryv1alpha1.ForeignCluster, gv *schema.GroupVersion) (*crdClient.CRDClient, error) {
+	if strings.HasPrefix(fc.Spec.AuthUrl, "fake://") {
+		config := *r.ForeignConfig
+
+		config.ContentConfig.GroupVersion = gv
+		config.APIPath = "/apis"
+		config.NegotiatedSerializer = client_scheme.Codecs.WithoutConversion()
+		config.UserAgent = rest.DefaultKubernetesUserAgent()
+
+		fc.Status.AuthStatus = discoveryv1alpha1.AuthStatusAccepted
+
+		return crdClient.NewFromConfig(&config)
+	}
+
+	if client, err := r.getIdentity(fc, gv); err == nil {
+		return client, nil
+	} else if !kerrors.IsNotFound(err) {
+		klog.Error(err)
+		return nil, err
+	}
+
+	if fc.Status.AuthStatus == discoveryv1alpha1.AuthStatusAccepted {
+		// TODO: handle this possibility
+		// this can happen if the role was accepted but the local secret has been removed
+		err := errors.New("auth status is accepted but there is no role ref")
+		klog.Error(err)
+		return nil, err
+	}
+
+	// not existing role
+	if fc.Status.AuthStatus == discoveryv1alpha1.AuthStatusPending || (fc.Status.AuthStatus == discoveryv1alpha1.AuthStatusEmptyRefused && r.getAuthToken(fc) != "") {
+		kubeconfigStr, err := r.askRemoteIdentity(fc)
+		if err != nil {
+			klog.Error(err)
+			return nil, err
+		}
+		if kubeconfigStr == "" {
+			return nil, nil
+		}
+		_, err = kubeconfig.CreateSecret(r.crdClient.Client(), r.Namespace, kubeconfigStr, map[string]string{
+			"cluster-id":           fc.Spec.ClusterIdentity.ClusterID,
```

Comment:
Please parameterize these labels and use a liqo prefix (e.g., `liqo.io/cluster-id`, `liqo.io/remote-identity`)